raft/server.hh

```diff
@@ -59,6 +59,13 @@ public:
     //
     // The caller may pass a pointer to an abort_source to make the operation abortable.
     // If abort is requested before the operation finishes, the future will contain `raft::request_aborted` exception.
+    //
+    // Successfull `add_entry` with `wait_type::committed` does not guarantee that `state_machine::apply` will be called
+    // locally for this entry. Between the commit and the application we may receive a snapshot containing this entry,
+    // so the state machine's state 'jumps' forward in time, skipping the entry application.
+    // However, for `wait_type::applied`, we guarantee that the entry will be applied locally with `state_machine::apply`.
```

Comment:
Do we? What we guaranty is that if you observe the state of your state machine after the function returns successfully it will include the application of your command. You are trying to make the guaranty much stronger. I am not sure this is a good idea since it may be a problem to preserve this stronger guaranty going forward.

Reply:
> Do we?

Yes, that's what the commit does

Reply:
My point is that before the commit there was no such guaranty and I am afraid it is too strict to introduce it.

Reply:
What's to fear? As you can see, the patch is very simple. The property is very easy to check: if apply index is `>=` our index, and snapshot index is `<`, then it **must** have been applied by `state_machine::apply`.
And we have nemesis test which fails without this guarantee.

Reply:
OTOH in the `drop_waiters` we do not check the term of the last entry in the snapshot so the change here aligns the behaviour of getting a snapshot before call to `wait_for_entry` and after.

Reply:
> What's to fear? As you can see, the patch is very simple. The property is very easy to check: if apply index is `>=` our index, and snapshot index is `<`, then it **must** have been applied by `state_machine::apply`. And we have nemesis test which fails without this guarantee.

The patch may be simple, but the guaranty is more strict. If we put the guaranty in the API it means that no matter how we change the implementation we have to preserve it.

Reply:
Well, I think it's a useful guarantee, users may want to use `state_machine::apply` to communicate with the caller of `add_entry` like I do inside nemesis test and like Tomek wanted to do when we imeplemented group 0 state machine. It's a natural thing.

Reply:
I am not saying it is not useful, but it may limit us in the future.

Reply:
Good, let it limit us, better have it explicitly stated and tested than people relying on implementation details for correctness and then having their code unexpectedly broken

Reply:
> > Then we will be able to mark all the waiters as succeeded.
> 
> Will be able, but we won't be forced to.
> 
> BTW. what you propose is already possible in some cases. When receiving a snapshot, compare the term at the snapshot's index, if we have that entry. If it's the same, we can mark all entries up to this index as applied.

May we should implement it then.

> 
> So we can do that, but is it actually useful? It doesn't seem to be. During normal operation (no failures), assuming that `snapshot_trailing` is large enough (and the default is `200`, so it's pretty large), followers won't get any snapshots, because they'll be able to keep up with the leader in appending and applying entries. So this "optimization" is only useful for failure scenarios. And the failure scenario where this "optimization" would change anything is one in which a follower appended 200 committed entries but did not manage to apply them before it received a later snapshot from a leader, which I think would be a pretty rare scenario.
>

A slow follower will benefit from the optimization. And there is always one slow follower the question is only if it is slow enough.
 
> I think that having the guarantee, which we both agree is useful, is better than being able to implement some optimizations of debatable use.

May be we should have three types of waits?

Reply:
> A slow follower will benefit from the optimization. And there is always one slow follower the question is only if it is slow enough.

If there was such a slow follower, which benefits from this particular optimization, that would imply that the follower would have to be regularly updated with a snapshot because it constantly would not be able to keep up with applying entries. This doesn't sound good, it sounds like we make the situation worse by forcing an already overloaded follower to become even more overloaded by having to download and apply snapshots all the time. I think in this situation, it would be better to slow down the entire system a bit (i.e. the leader would append or commit entries less frequently), otherwise we risk slowing it down _even more_ in an uncontrollable fashion.

> May be we should have three types of waits?

Sounds like an idea. It gives the users choice if they need that guarantee.

So here's my proposal:
1. go on with the PR, the current `wait_type::applied` will give the guarantee
2. if we want to introduce some optimizations in the future that break the guarantee, introduce a new `wait_type` for which the optimization works. That's a backward compatible change.

Reply:
> > A slow follower will benefit from the optimization. And there is always one slow follower the question is only if it is slow enough.
> 
> If there was such a slow follower, which benefits from this particular optimization, that would imply that the follower would have to be regularly updated with a snapshot because it constantly would not be able to keep up with applying entries. This doesn't sound good, it sounds like we make the situation worse by forcing an already overloaded follower to become even more overloaded by having to download and apply snapshots all the time. I think in this situation, it would be better to slow down the entire system a bit (i.e. the leader would append or commit entries less frequently), otherwise we risk slowing it down _even more_ in an uncontrollable fashion.

A snapshot may be much smaller then all the entries. Applying a snapshot may be much faster then applying 1000 entries.

When you have a system where a leader may proceed with sending more data without waiting for every node it is common to have one node that does not keeps up. We see it all the time with CL<ALL writes and had to invent a way to throttle writes when background work accumulates.  Transferring a snapshot is an alternative to throttling. You just drop all background work and replace it with one that is much lighter. Of course for state machines for which snapshot transfer is heavy there will have to be throttling.

> 
> > May be we should have three types of waits?
> 
> Sounds like an idea. It gives the users choice if they need that guarantee.
> 
> So here's my proposal:
> 
>     1. go on with the PR, the current `wait_type::applied` will give the guarantee
> 
>     2. if we want to introduce some optimizations in the future that break the guarantee, introduce a new `wait_type` for which the optimization works. That's a backward compatible change.

OK. But I like your idea to match a snapshot and mark all entries as applied. Just did not think about it when I wrote that code.

Reply:
I created an issue describing the optimization:
https://github.com/scylladb/scylla/issues/10695

Reply:
So if you agree with the plan, please give an LGTM so @tgrabiec can merge.